src/libraries/System.Linq/src/System/Linq/Chunk.cs

```diff
@@ -50,27 +51,24 @@ public static IEnumerable<TSource[]> Chunk<TSource>(this IEnumerable<TSource> so
         private static IEnumerable<TSource[]> ChunkIterator<TSource>(IEnumerable<TSource> source, int size)
         {
             using IEnumerator<TSource> e = source.GetEnumerator();
-            while (e.MoveNext())
-            {
-                TSource[] chunk = new TSource[size];
-                chunk[0] = e.Current;
 
-                int i = 1;
-                for (; i < chunk.Length && e.MoveNext(); i++)
+            if (e.MoveNext())
+            {
+                List<TSource> chunkBuilder = new();
+                do
                 {
-                    chunk[i] = e.Current;
-                }
+                    chunkBuilder.Clear();
+                    do
+                    {
+                        chunkBuilder.Add(e.Current);
+                    }
+                    while (chunkBuilder.Count < size && e.MoveNext());
 
-                if (i == chunk.Length)
-                {
-                    yield return chunk;
-                }
-                else
-                {
-                    Array.Resize(ref chunk, i);
+                    TSource[] chunk = new TSource[chunkBuilder.Count];
+                    chunkBuilder.CopyTo(chunk, 0);
```

Comment:
How is this different than `yield return chunkBuilder.ToArray()`?

Reply:
`ToArray()` for `List<T>` ends up calling `ICollection<T>.CopyTo()` after some casts, so essentially it's the same thing.

[in this comment](https://github.com/dotnet/runtime/issues/67132#issuecomment-1078951023) it felt like you were concerned with incurring that kind of extra cost so I was trying to keep in the spirit of that.

That said, in general the conversation around this has shifted away from micro-optimizing so if you'd prefer to save two lines of code that would also make sense to me. **Let me know and I'll update.**